csrc/ops/arith.cpp

```diff
@@ -2542,7 +2542,7 @@ TensorView* fusedMultiplySum(
     const std::vector<int>& axes,
     Val* init) {
   if (init == nullptr) {
-    init = IrBuilder::create<Val>(0.0);
+    init = IrBuilder::create<Val>(0.0, DataType::Float);
```

Comment:
Using incorrect dtype `Double` instead of `Float` caused the `init` value in the reduction to be printed as double, leading to a type error in compilation. This `init` variable seems to only be used by rfactor, so maybe we should just set it (properly) there instead of holding it as an `MmaOp` attribute. Would we ever use a non-zero `init` in an `MmaOp`?

Reply:
Can you show the code that caused the compilation error?

Reply:
Sure!
```c++
          reduction::serialReductionStep(                                                                                                              
            T6[i115],                                                                                                                                  
            T12[(i110 + i115)],                                                                                                                        
            0.00000000000000000e+00,                                                                                                                   
            T14[(i111 + (i115 + nvfuser_zero))],                                                                                                       
            [](float &a, float b) { a = a + b; },                                                                                                      
            index_utils::maskedOffset<false, false, true>(blockIdx, gridDim) == 0,                        
            index_utils::maskedOffset<false, false, true>(blockIdx, gridDim) == index_utils::maskedSize<false, false, true>(gridDim) - 1,              
            b114,                                                                                                                                      
            b114);                                                                                                                                     
        }                                                                                                                                              
```                                                               
The error is
```
CUDA NVRTC compile error: __tmp_kernel_none_f0_c0_r0_g0.cu(9806): error: no instance of function template "<unnamed>::reduction::serialReductionStep" matches the argument list
            argument types are: (float, float, double, float, lambda [](float &, float)->void, __nv_bool, __nv_bool, __nv_bool, __nv_bool)
            reduction::serialReductionStep(
```

Reply:
Changing init to `Float` makes it print properly:
```c++
          reduction::serialReductionStep(
            T6[i115],
            T12[(i110 + i115)],
            0.000000000e+00f,
            T14[(i111 + (i115 + nvfuser_zero))],
            [](float &a, float b) { a = a + b; },
            index_utils::maskedOffset<false, false, true>(blockIdx, gridDim) == 0,
            index_utils::maskedOffset<false, false, true>(blockIdx, gridDim) == index_utils::maskedSize<false, false, true>(gridDim) - 1,
            b114,
            b114);
```

Reply:
Looks like it's a template matching issue. Doesn't that mean the type just needs to be the same as the other parameters? That's most likely `float`, but theoretically could be other FP types as well, right?

Reply:
Oh good point, maybe I could cast `grop->init()` to `grop->in()->dtype()` instead possibly in `codegen.cpp`.

Reply:
Why not just using the dtype of `tv_a`?

Reply:
Ah, but `init` needs to be in the accumulator type. The inputs might be half precision.

Reply:
This function (`fusedMultiplySum`) calls `newForMma` to set up the output, so it is currently always defaulting to `Float`. It would be better to set up `init` after we get `out`, so that we can get the output type from there.

Reply:
I just pushed a change that uses `out->dtype()` as the dtype for `init`.